[Atoms] ListItem (ex MenuListItem)

nuintee opened the pull request:
## 概要
ListItemの実装 (元 MenuListItem PR)
`MenuListItem`と`CommonListItem`のコンポーネントが統合されたListItemになります。

<img width="1727" alt="スクリーンショット 2023-10-30 20 31 03" src="https://github.com/siva-squad/squad-ui/assets/61968429/b51380d0-45af-487f-8509-295bf89f3c3a">


### ToDo
- [x] tv化
- [x] dropdownの分岐・スタイリング
- [x] dropdownとheader時の型分岐
- [x] ナビゲーション対応

### Jira
https://siva-inc.atlassian.net/browse/RPD-18?atlOrigin=eyJpIjoiMmM4ZjBkZGM4YWY2NDg2ZGI1ZjU3YzdhYmU5ZTdlOGMiLCJwIjoiaiJ9

### Figma
https://www.figma.com/file/5KaykTqeXvtv3yGH9NV2Ol/Design-System?type=design&node-id=2679-13368&mode=design&t=tueEn5jUBHh6Cp20-4

### Note
#8 is merged with this PR and should be closed.

macfrei commented on src/components/molecules/MenuListItem/MenuListItem.tsx:
```diff
@@ -0,0 +1,44 @@
+      {icon && <span className={MENU_LIST_ITEM_ICON_CLASS_NAME({ selected })}>{iconUI}</span>}
+      <div className="flex flex-col items-start">
+        <span className={MENU_LIST_ITEM_TEXT_CLASS_NAME({ selected })}>{children}</span>
+        <span className={MENU_LIST_ITEM_DESCRIPTION_CLASS_NAME({ selected })}>{description}</span>
```
I think it would be better to use optional chaining to display the `description` or not:
```suggestion
        {description.length > 0 && <span className={MENU_LIST_ITEM_DESCRIPTION_CLASS_NAME({ selected })}>{description}</span>}
```
Otherwise the `span` will always render, whether there is a description or not.

macfrei commented on src/components/molecules/MenuListItem/type.ts:
```diff
@@ -0,0 +1,21 @@
+import { ReactElement, ReactNode } from "react";
+
+type TypeAndDescriptionType =
+  | {
+      type: "header";
+      description?: string;
+    }
+  | {
+      type: "dropdown";
+      description?: never;
+    };
+
+export type MenuListItemProps = {
+  chevron?: boolean;
+  disabled?: boolean;
+  icon?: ReactElement;
```
I think I might prefer combining these types:
```suggestion
export type MenuListItemProps = {
  type: "header" | "dropdown";
  description?: string;
  chevron?: boolean;
  disabled?: boolean;
  icon?: ReactElement;
```
At least to me that would be easier to read. What do you think?

macfrei reviewed (commented):
Added a few comments, please have a look 👍 🙇‍♀️

nuintee commented:
@macfrei 
I added breaking changes, due to the discussion with the designer. (I added navigation functionality)

I'd appreciate if you could review it again. 🙏

I'm still going to reply your comments that also exitsts in the new version.

nuintee commented on src/components/molecules/MenuListItem/type.ts:
```diff
@@ -0,0 +1,21 @@
+    }
+  | {
+      type: "dropdown";
+      description?: never;
```
Because of the Conditional Types

To make it accept description prop only when it's "header" type and not "dropdown"

nuintee commented on src/components/molecules/MenuListItem/MenuListItem.tsx:
```diff
@@ -0,0 +1,44 @@
+}: MenuListItemProps) => {
+  const iconUI = useMemo(() => {
+    if (selected) {
+      return selectedIcon || icon;
```
@macfrei 

At least, I thought so. 

Some of them could be just changing colors on selected state.

nuintee commented on src/components/molecules/MenuListItem/MenuListItem.tsx:
```diff
@@ -0,0 +1,44 @@
+  selectedIcon,
+  disabled,
+  selected = false,
+  onClick,
```
Thanks!

I'll change them!

nuintee commented on src/components/molecules/MenuListItem/type.ts:
```diff
@@ -0,0 +1,21 @@
+    };
+
+export type MenuListItemProps = {
+  chevron?: boolean;
+  disabled?: boolean;
+  icon?: ReactElement;
+  selectedIcon?: ReactElement;
+  children?: ReactNode;
+  selected?: boolean;
```
@macfrei 

Thanks!

I replaced `children` to `title` and made required following props.
- `title`
- `onClick` (when the type is `dropdown`)
- `href` (when the type is `header`)

### Reasons to make these props optional
- `icon` : not always given
- `selectedIcon`: not always given (At least I think so)
- `chevron`, `disabled`, `isSelected`: So that we don't have to pass `false` implicitly since we can pass nothing and still get the same result if it's optional.

nuintee commented:
@ogaru-sh
MenuListItemとCommonListItemがデザイン上で統合され一つのコンポーネントになったので、その対応をしました。

過去分のレビューを既に頂いておりましたが、諸々の修正が入っておりますのでもう一度レビューして頂けると幸いです。🙇‍♂️

\---

@macfrei 
Could you re review this PR? 🙏
I merged `MenuListItem` and `CommonListItem` components into this `ListItem` due to design file change.

ogaru-sh commented on src/components/atoms/ListItem/ListItem.tsx:
```diff
@@ -0,0 +1,61 @@
+        disabled={isDisabled}
+        onClick={onClick}
+        ref={ref as React.RefObject<HTMLAnchorElement> & React.RefObject<HTMLButtonElement>}
+        {...(href && { href })}
```
aタグのみかもしれないですが、`target`属性はあっても良いかもと思いました！どうでしょうか？ 
（とはいえ使い方によるかもしれない）

nuintee replied:
@ogaru-sh 
ありがとうございます！
確かにあっても良さそうですね。
追加します!

nuintee replied:
@ogaru-sh 
追加しました！

ogaru-sh replied:
ありがとうございます！
LGTM!

yoheeeei reviewed (dismissed):
LGTM 👍

macfrei reviewed (dismissed):
LGTM 🚀

nuintee commented:
@ogaru-sh 
デザイナーさんからもapprove頂けたので、お二方のどちらかにapproveステータスを頂けると有り難いです！

\---

@macfrei 
I have received approvals from the designers and would be grateful if either of you could give me an approve status!

nuintee commented:
@macfrei 
Thank you!
